File: java-practice/security/src/main/java/com/programmersstudy/springbootpractice/security_practice/config/security/config/SecurityConfig.java
package com.programmersstudy.springbootpractice.security_practice.config.security.config;

import com.programmersstudy.springbootpractice.security_practice.domain.Users;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.crypto.factory.PasswordEncoderFactories;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.provisioning.InMemoryUserDetailsManager;
import org.springframework.security.web.SecurityFilterChain;

@Configuration
public class SecurityConfig {
    private final PasswordEncoder passwordEncoder =
            PasswordEncoderFactories.createDelegatingPasswordEncoder();

    @Bean
    UserDetailsService authentication() {
        Users lkdcode = Users.builder()
                .email("lkdcode@email.com")
                .password(passwordEncoder.encode("password123"))
                .grade("GOLD")
                .role("ADMIN")
                .build();

        Users another = Users.builder()
                .email("another@email.com")
                .password(passwordEncoder.encode("password123"))
                .grade("SILVER")
                .role("ADMIN")
                .build();

        // TODO : BadCredentialsException handlingf
        System.out.println(lkdcode.getEmail() + " " + lkdcode.getPassword());
        System.out.println(another.getEmail() + " " + another.getPassword());
        return new InMemoryUserDetailsManager(lkdcode, another);
    }

    @Bean
    public SecurityFilterChain configure(HttpSecurity http) throws Exception {
        return http.authorizeHttpRequests()
                .antMatchers("/security/admin").hasAuthority("ADMIN")
                .antMatchers("/security/gold").hasAuthority("GOLD")
//                .requestMatchers("/security/admin").hasAuthority("ADMIN") Boot 3.0.2ver
//                .requestMatchers("/security/gold").hasAuthority("GOLD") Boot 3.0.2ver
                .anyRequest().authenticated()
                .and()
                .formLogin()
                .and()
                .httpBasic()
                .and()
                .build();
    }
}

File: java-practice/moku-demo/src/main/kotlin/run/moku/framework/config/SecurityConfig.kt
package run.moku.framework.config

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.ProviderManager
import org.springframework.security.authentication.dao.DaoAuthenticationProvider
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import run.moku.framework.security.service.MokuUserDetailsService

@Configuration
class SecurityConfig {
    @Bean
    fun bCryptPasswordEncoder(): PasswordEncoder = BCryptPasswordEncoder()

    @Bean
    fun authenticationManager(
        userDetailsService: MokuUserDetailsService,
        passwordEncoder: PasswordEncoder,
    ): AuthenticationManager {
        val authProvider = DaoAuthenticationProvider()
        authProvider.setUserDetailsService(userDetailsService)
        authProvider.setPasswordEncoder(passwordEncoder)
        return ProviderManager(authProvider)
    }
}